pallets/file-system/src/utils.rs

```diff
@@ -837,6 +814,135 @@ where
         Ok((bsp_id, new_root))
     }
 
+    pub(crate) fn do_delete_file(
+        sender: T::AccountId,
+        bucket_id: ProviderIdFor<T>,
+        file_key: MerkleHash<T>,
+        location: FileLocation<T>,
+        fingerprint: Fingerprint<T>,
+        size: StorageData<T>,
+        maybe_inclusion_forest_proof: Option<ForestProof<T>>,
+    ) -> Result<ProviderIdFor<T>, DispatchError> {
+        // Compute the file key hash.
+        let computed_file_key = Self::compute_file_key(
+            sender.clone(),
+            bucket_id,
+            location.clone(),
+            size,
+            fingerprint,
+        );
+
+        // Check that the metadata corresponds to the expected file key.
+        ensure!(
+            file_key == computed_file_key,
+            Error::<T>::InvalidFileKeyMetadata
+        );
+
+        let msp_id = <T::Providers as ReadProvidersInterface>::get_msp_of_bucket(&bucket_id)
+            .ok_or(Error::<T>::BucketNotFound)?;
+
+        match maybe_inclusion_forest_proof {
+            // If the user did not supply a proof of inclusion, queue a pending deletion file request.
+            // This will leave a window of time for the MSP to provide the proof of (non-)inclusion.
+            // If the proof is not provided within the TTL, the hook will queue a priority challenge to remove the file key from all the providers.
+            None => {
+                let pending_file_deletion_requests = <PendingFileDeletionRequests<T>>::get(&sender);
+
+                // Check if the file key is already in the pending deletion requests.
+                ensure!(
+                    !pending_file_deletion_requests
+                        .iter()
+                        .any(|(key, b_id)| key == &file_key && *b_id == bucket_id),
+                    Error::<T>::FileKeyAlreadyPendingDeletion
+                );
```

Comment:
File keys are unique for every single file, they include the bucket ID in their construction. So why not using `pending_file_deletion_requests.contains(file_key`?

Reply:
I changed it to `.contains(&(file_key.clone(), bucket_id.clone())),`. 

I don't want to remove the bucket id from storage because then users and providers would have to do perform extra queries to figure out which bucket the file key belongs to since you can't know the pre image without fetching it from some indexer.